File: app/src/main/java/com/example/dkalev/remember/model/CardViewModel.java
package com.example.dkalev.remember.model;

import android.arch.lifecycle.ViewModel;

import io.reactivex.Completable;
import io.reactivex.Flowable;

public class CardViewModel extends ViewModel {

    private final DeckDataSource mDataSource;
    private Card mCard;

    public CardViewModel(DeckDataSource dataSource){
        mDataSource = dataSource;
    }

    public Flowable<String> getTextFront(int card_uid){
        return mDataSource.getCard(card_uid)
                .map(card -> {
                    mCard = card;
                    return card.getTextFront();
                });
    }

    public Flowable<String> getTextBack(int card_uid){
        return mDataSource.getCard(card_uid)
                .map(card -> {
                    mCard = card;
                    return card.getTextBack();
                });
    }

    public Completable setTextFront(final String textFront){
        return Completable.fromAction(()->{
            Card card = new Card(mCard.getDeckId());
            card.setUid(mCard.getUid());
            card.setTextFront(textFront);
            card.setTextBack(mCard.getTextBack());
            mCard = card;
            mDataSource.updateCard(mCard);
        });
    }

    public Completable setTextBack(final String textBack){
        return Completable.fromAction(()->{
            Card card = new Card(mCard.getDeckId());
            card.setUid(mCard.getUid());
            card.setTextFront(mCard.getTextFront());
            card.setTextBack(textBack);
            mCard = card;
            mDataSource.updateCard(mCard);
        });
    }


    public Completable deleteCard(){
        return Completable.fromAction(() -> mDataSource.deleteCard(mCard));
    }
}

File: app/src/main/java/com/example/dkalev/remember/model/CardViewModel.kt
package com.example.dkalev.remember.model

import android.arch.lifecycle.ViewModel
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class CardViewModel(dataSource: DeckDataSource) : ViewModel() {

    private var mDataSource: DeckDataSource? = null
    private var mCard: Card? = null

    init {
        mDataSource = dataSource
    }

    private fun updateDB(func: () -> Unit): Completable {
        return Completable.fromAction{ func() }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    private fun fetchFromDB(func: () -> Flowable<String>): Flowable<String> {
        return func()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getTextFront(card_uid: Int): Flowable<String> {
        return fetchFromDB { mDataSource!!.getCard(card_uid)
                .map { card ->
                    mCard = card
                    card.textFront
                }
        }
    }

    fun getTextBack(card_uid: Int): Flowable<String> {
        return fetchFromDB {
            mDataSource!!.getCard(card_uid)
                    .map { card ->
                        mCard = card
                        card.textBack
                    }
        }
    }

    fun setTextFront(textFront: String): Completable {
        return updateDB {
            val card = Card(mCard!!.getDeckId())
            card.uid = mCard!!.getUid()
            card.textFront = textFront
            card.textBack = mCard!!.getTextBack()
            mCard = card
            mDataSource?.updateCard(mCard)
        }
    }

    fun setTextBack(textBack: String): Completable {
        return updateDB {
            val card = Card(mCard!!.getDeckId())
            card.uid = mCard!!.getUid()
            card.textFront = mCard!!.getTextFront()
            card.textBack = textBack
            mCard = card
            mDataSource?.updateCard(mCard)
        }
    }


    fun deleteCard(): Completable {
        return updateDB { mDataSource?.deleteCard(mCard) }
    }
}